File: problem01/twoSum.java
import java.util.Map;
import java.util.TreeMap;

/**
 * twoSum
 */
public class twoSum {

    public static void main(String[] args) {
        int[] nums = { -10, 7, 19, 15 };
        int target = 9;

        int[] result = twoSum2(nums, target);

        System.out.println("Index 1: " + result[0]);
        System.out.println("Index 2: " + result[1]);
    }

    private static int[] twoSum2(int[] nums, int target)
    {
        Map<Integer, Integer> dict = new TreeMap<>();
        int[] result = new int[2];

        for (int i = 0; i < nums.length; i++)
        {
            int index = target - nums[i];
            if (dict.containsKey(target - nums[i]) && i != dict.get(index))
            {
                result[0] = dict.get(index);
                result[1] = i;
                break;
            }
            else
            {
                dict.put(nums[i], i);
            }
        }
        return result;
    }
}

File: problem01/twoSum.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

class Solution
{
public:
    vector<int> twoSum(vector<int> &nums, int target)
    {
        std::unordered_map<int, int> newMap;
        vector<int> result;
        for (int i = 0; i < nums.size(); i++)
        {
            if (newMap.count(target - nums[i]) && i != newMap[target - nums[i]])
            {
                result.push_back(newMap[target - nums[i]]);
                result.push_back(i);
                break;
            }
            else
            {
                newMap[nums[i]] = i;
            }
        }
        return result;
    }
};

int main(int argc, char const *argv[])
{
    vector<int> nums;
    nums.push_back(-10);
    nums.push_back(7);
    nums.push_back(19);
    nums.push_back(15);
    int target = 9;
    Solution s;
    vector<int> result = s.twoSum(nums, target);
    cout << "Index 1: " << result[0] << endl;
    cout << "Index 2: " << result[1] << endl;
    return 0;
}
